### PR TITLE
Logplayer increasing delay fix

### PR DESCRIPTION
**Pull request by erniejunior**
Here we go. I just found a feature branch fixing #47 . Would be awesome if you could have a look at it so we get it upstream.

### CODE REVIEW

File: tools/cpp/logplayer/main.cpp [hunk 1]
```diff
@@ -270,46 +270,62 @@ struct LogPlayer
...
             if (diff > 3 && startedPub) nanosleep(&delay, nullptr);
 
+            
+            // busy waiting the rest
+            //std::cout << "prior: " << TimeUtil::utime()  << std::endl;
+            int64_t prior = TimeUtil::utime();
+            while(logDiffSpeed > TimeUtil::utime() - firstDispatchUtime){
+                //std::cout << "diff: " << logDiffSpeed - (TimeUtil::utime() - firstDispatchUtime) << std::endl;
+                }
+            std::cout << "timebusywait: " << (TimeUtil::utime() - prior) << std::endl;
```
Comment (erniejunior) [inline tools/cpp/logplayer/main.cpp]: TODO @erniejunior no std::cout should be left in this PR

**Comment by jbendes** [comment]
Hey @erniejunior, the zcm repo has whitespace trimmed at the end of each line. Looks like whatever editor you're using to work on this leaves whitespace (even on completely empty lines). Makes the diffs look like this:
image

Mind changing up your editor configuration to remove whitespace? Also this repo uses spaces (not tabs). Not sure which you're using

**Comment by jbendes** [comment]
Just pushed up some comments on https://github.com/ZeroCM/zcm/commit/2f641fcb2389697ee52d831c293317cfc078f019

Check out `nanosleep(...)` in the linux kernel and let me know what you think. http://man7.org/linux/man-pages/man2/nanosleep.2.html

**Comment by jbendes** [comment]
Also please rebase your branch on top of `master` before we merge

**Comment by erniejunior** [comment]
Thanks for the review. I merged with your comments branch and rebased onto master. Additinally I made some simplifications to the code.

Sorry for the unnecessary spaces. I do not know what editor was used to prepare the PR.

The [old resources about nanosleep](http://www.tldp.org/HOWTO/IO-Port-Programming-4.html) are unfortunately outdated. If you read the `notes` section of the manpage you suggested, you see that the busy-waiting feature of nanoseconds is not available any more since kernel 2.6. therefore wee need to do the busy waiting ourselves.

**Comment by jbendes** [comment]
@erniejunior Just took a look at your branch, rebased, and pushed up. Take a look when you get a chance. We're good to merge if you like what we did.

**Comment by jbendes** [comment]
As a heads up, we believe the uncertainty around `usleep` and `nanosleep` is in the single digit or low two digit millisecond range. So for a 10ms busy wait, and 1000Hz of data in a lot, logplayer will peg an entire CPU. For this reason we thought it best to release this functionality under a feature flag.

**Comment by jbendes** [comment]
Actually on second thought, we might just want to ditch the busy wait altogether. This branch (even with the busy wait turned off) seems to fix the inaccuracies in timing of playback. This is almost definitely due to your use of the firstDispatchUtime as opposed to the lastDispatchUtime that was being used before (which was probably accruing a bunch of error during runtime). I don't think anyone is going to use the busy wait functionality given that it consumes a CPU. That.....and it doesn't seem to provide any real benefit haha. I'd recommend deleting that part of this pullreq and leaving the part that fixes the whole problem haha

**Comment by erniejunior** [comment]
Good point to make the high-accuracy mode optional! I still think that there are use cases for a high-accuracy mode e.g. when we want to recreate effects like network jitter. When we introduce the option now we still have the option to implement it in a smarter way later (e.g. by making use of a real-time kernel). I would vote to merge it as is :+1:

**Pull request merged**